contracts/strategies/allocation/QFOffchainCalc.sol

```diff
@@ -0,0 +1,190 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+pragma solidity 0.8.19;
+
+import "@solady/auth/Ownable.sol";
+import "@openzeppelin-upgradeable/proxy/utils/Initializable.sol";
+
+import {IAllocationStrategy} from "../../interfaces/IAllocationStrategy.sol";
+import {Allo} from "../../core/Allo.sol";
+import {Metadata} from "../../core/libraries/Metadata.sol";
+
+/// @title QFOffchainCalc
+/// @notice A strategy that allows users to apply to a pool and vote on applications
+/// @dev This strategy is used for QF pools that have offchain calculations
+/// @author allo-team
+contract QFOffchainCalc is IAllocationStrategy, Ownable, Initializable {
+    uint256 public applicationStartTime;
+    uint256 public applicationEndTime;
+    uint256 public votingStartTime;
+    uint256 public votingEndTime;
+
+    bool public identityRequired;
+
+    uint256 private _index;
+
+    Allo allo;
+
+    error APPLICATIONS_NOT_OPEN();
+    error IDENTITY_REQUIRED();
+    error VOTING_NOT_OPEN();
+    error NOT_ALLO();
+
+    event ApplicationCreated(address indexed applicant, uint256 indexed applicationId);
+    event ApplicationStatusUpdated(address indexed applicant, uint256 indexed applicationId, Status status);
+    event ApplicationVoted(
+        address indexed voter, uint256 indexed applicationId, uint256 vote, address token, address recipient
+    );
+
+    /// @notice Struct to hold details of an application
+    struct Application {
+        bytes32 identityId;
+        address recipient;
+        uint256 applicationId;
+        uint256 amount;
+        uint256 votes;
+        Status status;
+    }
+
+    // Note: this is mapped to the Allo global status's in the mapping below.
+    /// @notice Enum for the local status of the application
+    enum Status {
+        PENDING,
+        ACCEPTED,
+        REJECTED,
+        REAPPLIED
+    }
+
+    /// @notice applicationId - Status
+    mapping(uint256 => ApplicationStatus) public applicationStatus;
+
+    /// @notice applicationId -> Application
+    mapping(uint256 => Application) public applications;
+
+    modifier onlyAllo() {
+        if (msg.sender != address(allo)) {
+            revert NOT_ALLO();
+        }
+        _;
+    }
+
+    /// @notice Initialize the contract
+    /// @param _allo The address of the Allo contract
+    /// @param _owner The address of the owner
+    function initialize(address _allo, address _owner) public initializer {
```

Comment:
@zobront Would it make sense to have this callable from `Allo.sol` on create and have every strategy implement this.
This would mean we pass custom bytes data to allow init to work as expected based on the strategy.

Reply:
Yep, I think that's a great idea.

Reply:
I added to a recent PR - Let me know what you think.